cclient/extended_api: iota_client_get_bundle and iota_client_traverse_bundle

ifullgaz opened the pull request:
iota_client_get_bundle and iota_client_traverse_bundle extended APIs

# Test Plan:
Usual test plan

ifullgaz committed:
cclient/extended_api: iota_client_get_bundle and iota_client_traverse_bundle

th0br0 commented on cclient/iota_client_extended_api.c:
```diff
@@ -249,3 +249,80 @@ retcode_t iota_client_get_account_data(iota_client_service_t const* const serv,
+    // Check that the first transaction we get is really a tail transaction
+    // Its index must be 0 and the list of transactions empty.
+    // It is an error if the first transaction has index > 0
+    if (!utarray_len(tx_objs) && transaction_current_index(tx)) {
```
`tx_objs` is not a utarray

ifullgaz replied:
tx_objs is a transaction_objs_t `typedef UT_array transaction_objs_t`.
Should it be another type?

th0br0 replied:
sorry, I overlooked your changed definition.
You'll still have to initialise it via `utarray_new` and provide an ICD though?
https://troydhanson.github.io/uthash/utarray.html

ifullgaz replied:
Yes, I thought that the caller would do that, unless you think it's more appropriate to do it in the function, in which case I'll change tx_objs to `transaction_objs_t **tx_objs`?

th0br0 commented on cclient/iota_client_extended_api.c:
```diff
@@ -249,3 +249,80 @@ retcode_t iota_client_get_account_data(iota_client_service_t const* const serv,
+
+  do {
+    // Create request and response to get_trytes API
+    get_trytes_req = get_trytes_req_new();
```
We should reuse these instead of reallocing?

ifullgaz replied:
Ah, yes

th0br0 reviewed (changes requested):
Also add a test case?

thibault-martinez commented on cclient/iota_client_extended_api.c:
```diff
@@ -249,3 +249,80 @@ retcode_t iota_client_get_account_data(iota_client_service_t const* const serv,
+      goto cleanup;
+    }
+    // Create a new transaction with the received trytes
+    tx = transaction_deserialize(tmp_trytes);
```
`transaction_deserialize` does a dynamic allocation. No need for that here ? Use a statically allocated transaction object + `transaction_deserialize_from_trits`

ifullgaz replied:
We potentially want to return several transactions

thibault-martinez replied:
But `tx` will be copied into the array right ? So no need to have it dynamic ?

thibault-martinez commented on cclient/iota_client_extended_api.c:
```diff
@@ -249,3 +249,80 @@ retcode_t iota_client_get_account_data(iota_client_service_t const* const serv,
+      ret_code = RC_CCLIENT_INVALID_TAIL_HASH;
+      goto cleanup;
+    }
+    // If the transaction is valid, add it to the transactino list
```
Typo

ifullgaz replied:
Ok

thibault-martinez commented on cclient/iota_client_extended_api.h:
```diff
@@ -21,9 +21,9 @@ typedef flex_hash_array_t address_list_t;
-typedef flex_hash_array_t transaction_objs_t;
 typedef flex_hash_array_t booleans_t;
 typedef flex_hash_array_t input_list_t;
+typedef UT_array transaction_objs_t;
```
We already have `typedef UT_array bundle_transactions_t;` and related methods in `common/model/bundle.h`. Would it make sense, for semantic, to use them ?

oopsmonk replied:
Does `bundle_transactions_t` equal to a list of transaction object? IMO, it has different meaning between a bundle and a transactions list.

thibault-martinez replied:
`bundle_transactions_t` is an array of `iota_transaction_t`, the exact same thing as `transaction_objs_t`. But since here we are dealing with bundles maybe we should use it ?

oopsmonk replied:
ah..right, it is a bundle operation. sorry.

ifullgaz committed:
cclient/extended_api: iota_client_get_bundle, iota_client_traverse_bundle, iota_client_broadcast_bundle
utils/containers: added hash_stack_at function
common/model: replaced numeric constant with define

ifullgaz committed:
cclient/extended_api: iota_client_promote_transaction (in progress)

ifullgaz committed:
cclient/extended_api: iota_client_replay_bundle (in progress)

Commit:
cclient: fixing compile error.

Commit:
cclient: fixing memory leaks.

oopsmonk commented:
marked as not merger due to APIs still WIP, including `iota_client_get_bundle`, `iota_client_promote_transaction`, `iota_client_replay_bundle`

Commit:
cclient: fixing get_bundle & replay_bundle.

oopsmonk commented:
TODO `iota_client_promote_transaction`

th0br0 commented on cclient/iota_client_extended_api.c:
```diff
@@ -56,6 +56,148 @@ static retcode_t is_unused_address(iota_client_service_t const* const serv,
   return ret_code;
 }
 
+static retcode_t check_consistency(iota_client_service_t const* const serv,
```
this method should support multiple tails

th0br0 commented on cclient/iota_client_extended_api.c:
```diff
@@ -56,6 +56,148 @@ static retcode_t is_unused_address(iota_client_service_t const* const serv,
+
+cleanup:
+  check_consistency_req_free(&consistency_req);
+  check_consistency_res_free(consistency_res);
```
missing a &?

th0br0 commented on cclient/iota_client_extended_api.c:
```diff
@@ -56,6 +56,148 @@ static retcode_t is_unused_address(iota_client_service_t const* const serv,
+  ret_code =
+      iota_client_check_consistency(serv, consistency_req, consistency_res);
+  if (ret_code) {
+    log_warning(logger_id, "check consistency failed\n");
```
also log ret_code?

th0br0 commented on cclient/iota_client_extended_api.c:
```diff
@@ -56,6 +56,148 @@ static retcode_t is_unused_address(iota_client_service_t const* const serv,
+    transaction_deserialize_from_trits(&tx, tmp_trytes, true);
+
+    current_index = transaction_current_index(&tx);
+    last_index = transaction_last_index(&tx);
```
we should only read the last_index once (for first tx in bundle)

oopsmonk replied:
considering the logic here, it won't cause a problem and keep the code simple.

th0br0 commented on cclient/iota_client_extended_api.c:
```diff
@@ -56,6 +56,148 @@ static retcode_t is_unused_address(iota_client_service_t const* const serv,
+    // Create a new transaction with the received trytes
+    transaction_deserialize_from_trits(&tx, tmp_trytes, true);
+
+    current_index = transaction_current_index(&tx);
```
should probably make sure that we start at current_index = 0?

oopsmonk replied:
check tail at line#167

Commit:
cclient: removing private check_consistency method.

cclient: fixing test cases.

oopsmonk commented:
fixing #240, #252, #238 
let's do `iota_client_promote_transaction` in another PR.

thibault-martinez commented on common/model/transaction.c:
```diff
@@ -351,6 +352,28 @@ size_t transaction_deserialize_from_trits(iota_transaction_t *const transaction,
+                         NUM_TRYTES_ADDRESS);
+  flex_trits_from_trytes(transaction_tag(spam_transaction), NUM_TRITS_TAG,
+                         message, NUM_TRITS_TAG, NUM_TRYTES_TAG);
+  return spam_transaction;
```
`transaction_new()` already sets these fields to `FLEX_TRIT_NULL_VALUE`

oopsmonk replied:
missed this method, let's remove it.

thibault-martinez commented on cclient/iota_client_extended_api.c:
```diff
@@ -56,6 +56,108 @@ static retcode_t is_unused_address(iota_client_service_t const* const serv,
+  if (trytes) {
+    hash_array_free(trytes);
+  }
+  // done:
```
remove ?

thibault-martinez commented on cclient/iota_client_extended_api.c:
```diff
@@ -56,6 +56,108 @@ static retcode_t is_unused_address(iota_client_service_t const* const serv,
+                error_2_string(ret_code));
+      goto cleanup;
+    }
+    hash243_queue_free(&get_trytes_req->hashes);
```
`hash243_queue_pop` would achieve the same with less variables

thibault-martinez commented on cclient/iota_client_extended_api.c:
```diff
@@ -56,6 +56,108 @@ static retcode_t is_unused_address(iota_client_service_t const* const serv,
+    }
+    hash243_queue_free(&get_trytes_req->hashes);
+    // Get the transaction trytes
+    tmp_trytes = hash8019_queue_at(&get_trytes_res->trytes, 0);
```
Same here with `hash8019_queue_peek`

thibault-martinez commented on cclient/iota_client_extended_api.c:
```diff
@@ -56,6 +56,108 @@ static retcode_t is_unused_address(iota_client_service_t const* const serv,
+      goto cleanup;
+    }
+    // Create a new transaction with the received trytes
+    transaction_deserialize_from_trits(&tx, tmp_trytes, true);
```
We can avoid hashing since we already have the hash, we just need to set it

thibault-martinez commented on cclient/iota_client_extended_api.c:
```diff
@@ -56,6 +56,108 @@ static retcode_t is_unused_address(iota_client_service_t const* const serv,
+    current_index = transaction_current_index(&tx);
+    last_index = transaction_last_index(&tx);
+    // Get the next tail_hash from the transaction
+    memcpy(trunk_hash, transaction_trunk(&tx), FLEX_TRIT_SIZE_243);
```
Can't `trunk_hash` just be a pointer to avoid copies ?

oopsmonk replied:
I don't think so.

thibault-martinez commented on cclient/iota_client_extended_api.c:
```diff
@@ -56,6 +56,108 @@ static retcode_t is_unused_address(iota_client_service_t const* const serv,
+      bundle_transactions_add(bundle, &tx);
+    }
+    // The response is not needed anymore
+    hash8019_queue_free(&get_trytes_res->trytes);
```
`hash8019_queue_pop` too

thibault-martinez commented on cclient/iota_client_extended_api.c:
```diff
@@ -546,3 +645,148 @@ retcode_t iota_client_send_transfer(
+    goto cleanup;
+  }
+
+  log_debug(logger_id, "[%s:%d]\n", __func__, __LINE__);
```
What are these for ? Can we remove them ?

thibault-martinez commented on cclient/iota_client_extended_api.c:
```diff
@@ -546,3 +645,148 @@ retcode_t iota_client_send_transfer(
+  if (iota_client_traverse_bundle(serv, tail_hash, bundle) == RC_OK) {
+    bundle_validator(bundle, &bundle_status);
+  }
+  return bundle_status;
```
Should this be a parameter ? We are dismissing the real error

thibault-martinez commented on cclient/iota_client_extended_api.c:
```diff
@@ -546,3 +645,148 @@ retcode_t iota_client_send_transfer(
+
+  if (consistency_res->state) {
+    // adding spam transaction
+    spam_transaction = transaction_new();
```
Just noting for the next PR that this need to be freed

oopsmonk replied:
right, thanks.

Commit:
cclient: refactoring traverse_bundle.

Commit:
cclient: simplify traverse_bundle.

thibault-martinez commented on cclient/serialization/tests/test_json.c:
```diff
@@ -584,21 +589,23 @@ void test_serialize_attach_to_tangle(void) {
       (const tryte_t*)TEST_2673_TRYTES_1, NUM_TRYTES_SERIALIZED_TRANSACTION,
       NUM_TRYTES_SERIALIZED_TRANSACTION));
-  hash_array_push(attach_req->trytes, trits_8019);
+  hash_array_push(trytes, trits_8019);
```
I really think we should offer API functions like `attach_to_tangle_req_add_trytes` for this kind of operations. The end user is not supposed to know the underlying types and their functions.

oopsmonk replied:
good point, see issue #805

th0br0 commented on cclient/iota_client_extended_api.c:
```diff
@@ -56,6 +56,107 @@ static retcode_t is_unused_address(iota_client_service_t const* const serv,
+  }
+
+  do {
+    memcpy(trunk_hash, is_tail ? tail_hash : transaction_trunk(&tx),
```
we don't need the local `trunk_hash` here, can simply call `transaction_set_trunk(tx, tail_hash)` before the loop and always refer to the hash inside the loop via `transaction_trunk`

oopsmonk replied:
`trunk_hash` log the bundle hash for next transaction that we don't recalculate it during deserialization.

th0br0 replied:
no, it holds the tail tx's hash (i.e. H(tx in bundle with curIdx=0), ...)

oopsmonk replied:
done.

th0br0 commented on cclient/iota_client_extended_api.c:
```diff
@@ -56,6 +56,107 @@ static retcode_t is_unused_address(iota_client_service_t const* const serv,
+    }
+    // The response is not needed anymore
+    hash8019_queue_free(&get_trytes_res->trytes);
+  } while (current_index != last_index);
```
Doesn't the JS implementation do additional checks here?
What if the bundle we're trying to traverse is malformed maliciously? e.g. varying indices, ...
Ideally, we should at least check that the `current_index` of the retrieved transaction matches our expected index (`current_index+1`) and maybe also check that bundle hash is consistent

oopsmonk replied:
yes, JS doesn't have additional checks.
I expect the current_index is in the right order(0,1,2..) and the bundle hash is consistent since transactions are fetched from the Tangle.

th0br0 replied:
getTrytes doesn't do any server-side tx validation iirc

oopsmonk replied:
done.

Commit:
cclient: bundle checking in traverse_bundle.

Commit:
cclient: escaping is missing.

Commit:
Merge branch 'develop' into cclient_extended_api

Commit:
Merge branch 'develop' into cclient_extended_api

thibault-martinez committed:
Merge branch 'develop' into cclient_extended_api
